google-cloud-pubsublite/src/test/java/com/google/cloud/pubsublite/internal/BlockingPullSubscriberImplTest.java

```diff
@@ -0,0 +1,233 @@
+    SequencedMessage message =
+        SequencedMessage.of(Message.builder().build(), Timestamps.EPOCH, Offset.of(12), 30);
+    Future<?> future = executorService.submit(() -> subscriber.onData().get());
+    messageConsumer.accept(ImmutableList.of(message));
+    future.get();
+  }
+
+  @Test
+  public void onDataMultiCalls() {
+    Future<?> future1 = subscriber.onData();
+    Future<?> future2 = subscriber.onData();
+    ExecutionException e = assertThrows(ExecutionException.class, () -> future1.get());
+    assertThat(e.getCause()).isInstanceOf(InterruptedException.class);
+  }
+
+  @Test
+  public void pullMessage() throws Exception {
+    SequencedMessage message =
+        SequencedMessage.of(Message.builder().build(), Timestamps.EPOCH, Offset.of(12), 30);
+    messageConsumer.accept(ImmutableList.of(message));
+    assertThat(Optional.of(message)).isEqualTo(subscriber.messageIfAvailable());
+  }
+
+  @Test
+  public void pullMessageNoMessage() throws Exception {
+    assertThat(Optional.empty()).isEqualTo(subscriber.messageIfAvailable());
+  }
+
+  @Test
+  public void pullMessageWhenError() {
+    CheckedApiException expected = new CheckedApiException(StatusCode.Code.INTERNAL);
+    errorListener.failed(null, expected);
+    try {
+      subscriber.messageIfAvailable();
+      fail();
+    } catch (CheckedApiException e) {
+      assertThat(expected).isEqualTo(e);
+    }
+  }
+
+  @Test
+  public void pullMessagePrioritizeErrorOverExistingMessage() {
+    CheckedApiException expected = new CheckedApiException(StatusCode.Code.INTERNAL);
+    errorListener.failed(null, expected);
+    SequencedMessage message =
+        SequencedMessage.of(Message.builder().build(), Timestamps.EPOCH, Offset.of(12), 30);
+    messageConsumer.accept(ImmutableList.of(message));
+
+    try {
+      subscriber.messageIfAvailable();
+      fail();
+    } catch (CheckedApiException e) {
+      assertThat(expected).isEqualTo(e);
+    }
+  }
+
+  @Test
+  public void onlyOneMessageDeliveredWhenMultiCalls() throws Exception {
+    SequencedMessage message =
+        SequencedMessage.of(Message.builder().build(), Timestamps.EPOCH, Offset.of(12), 30);
+    messageConsumer.accept(ImmutableList.of(message));
+
+    AtomicInteger count = new AtomicInteger(0);
+    CountDownLatch latch = new CountDownLatch(1);
+    List<Future<?>> futures = new ArrayList<>();
+    for (int i = 0; i < 10; i++) {
+      futures.add(
+          executorService.submit(
+              () -> {
+                try {
+                  latch.await();
+                  if (subscriber.messageIfAvailable().isPresent()) {
+                    count.incrementAndGet();
+                  }
+                } catch (Exception e) {
+                  throw new IllegalStateException(e);
+                }
+              }));
+    }
+    latch.countDown();
+    for (Future<?> f : futures) {
+      f.get();
+    }
+    assertThat(1).isEqualTo(count.get());
+  }
```

Comment:
add a comment that this is not guaranteed to fail if the subscriber is not thread safe so if it is flaky, don't ignore it.